crates/natrix/src/reactivity/signal_list.rs

```diff
@@ -0,0 +1,563 @@
+    /// Create a new empty `SignalList`
+    #[inline]
+    #[must_use]
+    pub fn new() -> Self {
+        Self {
+            items: Vec::new(),
+            deps: RefCell::new(Vec::new()),
+            len_deps: RefCell::new(SignalDepList::new()),
+        }
+    }
+
+    /// Create a new `SignalList` with the specified capacity
+    #[inline]
+    #[must_use]
+    pub fn with_capacity(capacity: usize) -> Self {
+        Self {
+            items: Vec::with_capacity(capacity),
+            deps: RefCell::new(Vec::with_capacity(capacity)),
+            len_deps: RefCell::new(SignalDepList::new()),
+        }
+    }
+
+    /// Get the length of the list.
+    /// This subscribes to length changes (push/pop/clear).
+    #[inline]
+    #[must_use]
+    pub fn len(&self) -> usize {
+        // Track length dependency
+        if let Some(hook) = core::statics::current_hook() {
+            if let Ok(mut len_deps) = self.len_deps.try_borrow_mut() {
+                len_deps.insert(hook);
+            } else {
+                log_or_panic!("Length deps list already borrowed");
+            }
+        }
+        self.items.len()
+    }
+
+    /// Check if the list is empty.
+    /// This subscribes to length changes.
+    #[inline]
+    #[must_use]
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    /// Push an element to the end of the list.
+    /// This only invalidates length readers, not element readers.
+    pub fn push(&mut self, value: T) {
+        self.items.push(value);
+        
+        // Ensure deps vec is large enough
+        if let Ok(mut deps) = self.deps.try_borrow_mut() {
+            deps.push(SignalDepList::new());
+        } else {
+            log_or_panic!("Deps list already borrowed during push");
+        }
+
+        // Only invalidate length subscribers
+        core::statics::reg_dirty_list(|| {
+            self.len_deps.get_mut().create_iter_and_clear()
+        });
+    }
+
+    /// Pop an element from the end of the list.
+    /// This invalidates length readers and the last element's readers.
+    pub fn pop(&mut self) -> Option<T> {
+        let result = self.items.pop();
+        
+        if result.is_some() {
+            // Remove the dep list for the removed element and invalidate it
+            if let Ok(mut deps) = self.deps.try_borrow_mut() {
+                if let Some(mut last_deps) = deps.pop() {
+                    // Invalidate both the removed element and length
+                    core::statics::reg_dirty_list(|| last_deps.create_iter_and_clear());
+                    core::statics::reg_dirty_list(|| {
+                        self.len_deps.get_mut().create_iter_and_clear()
+                    });
+                } else {
+                    // Only invalidate length if no element deps to remove
+                    core::statics::reg_dirty_list(|| {
+                        self.len_deps.get_mut().create_iter_and_clear()
+                    });
+                }
```

Comment:
You cold just move the len one to outside the condition.


```suggestion
                if let Some(mut last_deps) = deps.pop() {
                    core::statics::reg_dirty_list(|| last_deps.create_iter_and_clear());
                }
                core::statics::reg_dirty_list(|| {
                    self.len_deps.get_mut().create_iter_and_clear()
                });
```